Handle bytecode without IC state

bnoordhuis opened the pull request:
I split this into three commits. The second one segfaults, the third one fixes that.

Interestingly, starting qjs in REPL mode segfaulted but the other bytecode tests we have don't. Not quite sure what's up with that but `bc->ic` not pointing to anything was definitely the issue though.

I added two (for now undocumented) options to `std.evalScript` that may be useful in their own right.

bnoordhuis commented:
Interesting...

> /Users/runner/work/quickjs/quickjs/quickjs.c:35019:21: runtime error: applying non-zero offset 27 to null pointer

I do believe ubsan is unearthing a (pre-existing) bug here.

saghul reviewed (approved):
LGTM! I also like you added those flags, I was planning on adding them myself to implement standalone executables in JS :-P

bnoordhuis commented:
I'll squash the 2nd and 3rd commit (don't want to break `git bisect`) and land this as three commits.

bnoordhuis committed:
Handle bytecode without IC state

Deserialized bytecode does not have IC state, i.e., `bc->ic == NULL`.
That may or may not be bug (IMO, it is and we should rebuild the
IC state during deserialization) but, either way, don't segfault.

DRY add_ic_slot() and its call sites in a hopefully NFC manner.

bnoordhuis committed:
Improve deserializer error message for bytecode

Don't raise a "invalid tag 12" exception when encountering bytecode
and JS_READ_OBJ_BYTECODE is not set, because no one knows what "tag 12"
means without looking it up, not even quickjs maintainers.

bnoordhuis committed:
Fix UndefinedBehaviorSanitizer error

UBSan is right to complain that `s->ptr_last == NULL` when tracing is
disabled.
